crates/provider/src/chain/stream.rs

```diff
@@ -0,0 +1,159 @@
+                }
+            }
+        }
+    }
+
+    async fn new(
+        weak_client: WeakClient<T>,
+        next_yield: Option<BlockNumber>,
+    ) -> Result<Self, ChainStreamError<T>> {
+        let client = Self::client(&weak_client)?;
+        let chain_tip = client
+            .request("eth_blockNumber", ())
+            .await
+            .map(|b: U64| b.to::<u64>())
+            .map_err(|e| ChainStreamError::Rpc(e))?;
+        let next_yield = next_yield.unwrap_or(chain_tip);
+
+        let block_numbers_stream = PollerBuilder::new(weak_client.clone(), "eth_blockNumber", ())
+            .spawn()
+            .into_stream_raw();
+        Ok(Self {
+            client: weak_client,
+            next_yield,
+            chain_tip,
+            block_numbers_stream,
+            _phantom: PhantomData,
+        })
+    }
+
+    fn client(client: &WeakClient<T>) -> Result<Arc<RpcClientInner<T>>, ChainStreamError<T>> {
+        client.upgrade().ok_or(ChainStreamError::ClientDropped)
+    }
+
+    /// Updates the chain tip so that it is greater than `next_yield`.
+    async fn update_chain_tip(&mut self) -> Result<(), ChainStreamError<T>> {
+        // Loop until chain tip is advanced.
+        while self.next_yield > self.chain_tip {
+            let new_number = match self.block_numbers_stream.next().await {
+                Some(Ok(block_number)) => block_number.to::<u64>(),
+                Some(Err(err)) => {
+                    // This is fine.
+                    debug!(%err, "polling stream lagged");
+                    continue;
+                }
+                None => {
+                    debug!("polling stream ended");
+                    return Err(ChainStreamError::PollingStreamEnded);
+                }
+            };
+            self.chain_tip = new_number;
+        }
+        Ok(())
+    }
+
+    /// Tries to fetch the range of blocks that can surely be yielded via single batch request.
+    async fn fetch_blocks(&self) -> Result<Vec<Block>, ChainStreamError<T>> {
+        // We don't want to request too many batches by accident.
+        // TODO: Should be a property on transport or client.
+        const MAX_BATCH_SIZE: usize = 10;
+
+        let range_end = self.chain_tip.min(self.next_yield + MAX_BATCH_SIZE as u64 - 1);
+        let range = self.next_yield..=range_end;
+        if range.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Perform a batch request.
+        let client = Self::client(&self.client)?;
+        let mut batch_request = BatchRequest::new(client.as_ref());
+        let futures: FuturesOrdered<Waiter<Option<Block>>> = range
+            .map(|number| {
+                batch_request
+                    .add_call::<_, Option<Block>>(
+                        "eth_getBlockByNumber",
+                        &(U64::from(number), false),
+                    )
+                    .expect("Cannot serialize call params")
+            })
+            .collect();
```

Comment:
I'm actually not sure how batch request works under the hood, but I believe there's no guarantee that responses are in the same order as requested

Reply:
Each request is represented by an individual future, so what we collect here is futures for responses we've created, so I think it's safe to assume that the order will be the same.
I've added a comment for that.